### ISSUE TITLE
Printing or asserting in device-side code can reduce or prevent concurrency

### ISSUE THREAD

**Issue by mruberry** [pytorch/pytorch#9123]
The problem:

- PyTorch has asserts in its kernels. These asserts, even if not triggered, can prevent these kernels from running concurrently with other kernels.

The severity of this issue depends on the CUDA driver, but all recent drivers have some performance degradation when attempting to use kernels with asserts (or prints) in them concurrently. 

In addition to the performance hit, the unexpected failure to run kernels concurrently may cause serious user and developer headaches.

Possible solutions:

- Remove asserts and prints from device code.
- Only enable asserts and prints for special debug builds.

**Comment by apaszke** [comment]
Well the asserts usually guard out-of-bounds accesses, and I don't know how we could solve it otherwise. I don't think silently ignoring them is a good idea, because this can lead to values from uninitialized memory creeping into someone's computation, and blowing up things like BatchNorm stats in the middle of the training. We discussed the performance hit, and concluded that it's not big enough to warrant the extra debugging difficulty, or silencing the errors.

I guess I would be ok with having two sets of kernels with asserts on and off, but I'm confident that the default should be on, and off would be a high-perf-I-know-what-I'm-doing mode.

**Comment by mruberry** [comment]
Power users can already disable asserts in kernels by building PyTorch with the correct flag, so we already have the specialized build we're talking about available to them.

I'll close this for now until a better alternative to asserts is developed.
